airflow/models/taskinstance.py

```diff
@@ -181,6 +182,7 @@ def clear_task_instances(
     task_id_by_key: dict[str, dict[str, dict[int, dict[int, set[str]]]]] = defaultdict(
         lambda: defaultdict(lambda: defaultdict(lambda: defaultdict(set)))
     )
+    dag_bag = DagBag(read_dags_from_db=True)
```

Comment:
I wonder why `dag` is not a required argument to begin with. It really should be.

Reply:
```ti_dag = dag if dag and dag.dag_id == ti.dag_id else dag_bag.get_dag(ti.dag_id, session=session)``` 

Looks like dag is optional because if there is no DAG we can just get the DAG from the TaskInstance

Reply:
With this change yes, but without this PR `dag=None` would have a very weird behaviour and not particularly useful. This makes me think the signature may be considered a bug in the first place and fixed without introducing `DagBag` here.